### ISSUE TITLE
Regression: Segfault when project has too many classes

### ISSUE THREAD

**Issue by spditner** [sorbet/sorbet#6805]
#### Input

Generate a file with many classes (or many files with a single class each), 32,500 will tip it over the threshold: 

```bash
echo > output.rb ; i=0; 
while [ $i -lt 32500 ]; do i=$((i+1)); echo "class MyClass$i; end" >> output.rb; done
```

Then run:

```
srb tc --no-config .
```

#### Observed output

```
~/.rbenv/versions/3.2.1/lib/ruby/gems/3.2.0/gems/sorbet-0.5.10693/bin/srb: line 46: 66439 Segmentation fault: 11  "${sorbet}" "${args[@]}"
```

#### Expected behavior

No segfault.

#### Additional background

* Issue introduced in https://github.com/sorbet/sorbet/pull/6763
* Fix attempted in https://github.com/sorbet/sorbet/pull/6792
* Tested above snippets with 0.5.10685 with up to 1,000,000 classes without issue.

**Comment by jez** [comment]
I'll have to look into this. I don't see an obvious cause, and it's tricky for us to revert those two PRs now.

**Comment by ashkulz** [comment]
I'm not sure if it's related, but using `srb rbi gems` on the `ruby_parser` gem (version 3.20) now generates a lot of methods which causes a segfault:
```
$ srb tc
Exception::raise(): isModule() but neither flags.isModule nor flags.isClass. Call to isModule before resolver?

Backtrace:
  #3 0x950127 sorbet::namer::(anonymous namespace)::defineSymbols()
  #4 0x94dfdb sorbet::namer::Namer::runInternal()
  #5 0x9559f3 sorbet::namer::Namer::run()
  #6 0x7c0369 sorbet::realmain::pipeline::name()
  #7 0x7c05bd sorbet::realmain::pipeline::resolve()
  #8 0x68cb32 sorbet::realmain::realmain()
  #9 0x6882dc main
  #10 0x7fb49da65d0a __libc_start_main
  #11 0x6423aa _start

/usr/local/rvm/gems/default/gems/sorbet-0.5.10696/bin/srb: line 46:  3678 Segmentation fault      (core dumped) "${sorbet}" "${args[@]}"
```
See the attached [ruby_parser.rbi](https://github.com/sorbet/sorbet/files/10894974/ruby_parser.rbi.txt).

**Comment by jez** [comment]
@spditner I reverted the revert, which originally closed this issue, and have rolled forward with what I believe should be a long term fix in #6883.

If you see the issue re-occur, please feel free to open a new issue.
